### PR TITLE
LEO-129: Add a Python 3 Kernel

### PR DESCRIPTION
**Pull request by ansingh7115**
#195 Add a python 3 kernel. 

Since this is modifying the Dockerfile, I created a fc-develop branch of the same name and made the jupyter docker image tag point to my branch name and not dev, so automated tests run against that image tag.

This PR does not include the upgrade to Debian 9 "stretch" - we can discuss the order of merging things.

The jupyter image has increased in size from 1.49GB to 2.09GB

Tech doc with more details:
 https://docs.google.com/document/d/1aGFcuZKzSZUMVINDI08_jeoQ3vJ09CCq5NyLNQ5bZYM

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

**Comment by ansingh7115** [comment]
jenkins retest

**Comment by ansingh7115** [comment]
Tests passed, but on separate runs

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/NotebooksListPage.scala [hunk 1]
```diff
@@ -7,6 +7,26 @@ import org.openqa.selenium.WebDriver
...
+  def string: String = "Python 2"
+  def cssSelectorString: String = "[title='Create a new notebook with Python 2']"
+
+}
+object Python3 extends Kernel {
+  def string: String = "Python 3"
+  def cssSelectorString: String = "[title='Create a new notebook with Python 3']"
+}
+
+object RKernel extends Kernel {
```
Comment (ansingh7115) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/NotebooksListPage.scala]: in anticipation for R kernel tests

File: jupyter-docker/Dockerfile [hunk 2]
```diff
@@ -127,6 +127,38 @@ RUN apt-get install -yq --no-install-recommends \
  && pip install matplotlib-venn \
  && pip install "https://github.com/all-of-us/workbench/archive/pyclient-v0-1-rc3.zip#egg=aou_workbench_client&subdirectory=client/py"
 
+
+# Install Python 3 Kernel
+RUN apt-get update \
+ && apt-get install -yq --no-install-recommends python3 \
+        liblzo2-dev \
```
Comment (ansingh7115) [inline jupyter-docker/Dockerfile]: needed this for bx-python package.

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/NotebooksListPage.scala [hunk 3]
```diff
@@ -7,6 +7,26 @@ import org.openqa.selenium.WebDriver
 
 import scala.util.Try
 
+trait Kernel {
```
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/NotebooksListPage.scala]: Minor, maybe make these `sealed trait` and `case object`s. That is useful if we're pattern matching on these things (which we're not currently doing, but still good practice).

File: jupyter-docker/Dockerfile [hunk 4]
```diff
@@ -127,6 +127,38 @@ RUN apt-get install -yq --no-install-recommends \
  && pip install matplotlib-venn \
  && pip install "https://github.com/all-of-us/workbench/archive/pyclient-v0-1-rc3.zip#egg=aou_workbench_client&subdirectory=client/py"
 
+
+# Install Python 3 Kernel
+RUN apt-get update \
```
Comment (rtitle) [inline jupyter-docker/Dockerfile]: Minor, is `apt-get update` needed given that we already did it above?

Comment (rtitle) [reply jupyter-docker/Dockerfile]: Oh, it looks like the previous layer cleans up apt caches:
```
&& apt-get clean \
&& rm -rf /var/lib/apt/lists/* \
```

Maybe remove that above, and only do that after all the `apt-get`s are done.

File: jupyter-docker/Dockerfile [hunk 5]
```diff
@@ -127,6 +127,38 @@ RUN apt-get install -yq --no-install-recommends \
...
+        python3-dev \
+        python3-matplotlib \
+        python3-pandas \
+        python3-tk \
+        python3-numpy \
+ && python3 get-pip.py \
+ && pip3 install ipykernel \
+ && python3 -m ipykernel install --name python3 --display-name "Python 3"
+
+RUN pip3 install seaborn \
```
Comment (rtitle) [inline jupyter-docker/Dockerfile]: Python 2 does the `apt-get`s and `pip install`s in the same layer. Could we do the same for py3?

I'm wondering if we should have 1 giant layer for py2 and py3; or 1 layer for py2 and 1 layer for py3. I'm leaning towards the latter.

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/NotebookInteractionSpec.scala [hunk 6]
```diff
@@ -175,5 +175,22 @@ class NotebookInteractionSpec extends FreeSpec with LeonardoTestUtils with Befor
       }
     }
 
+    "should create a notebook with a working Python 3 kernel and import installed packages" in withWebDriver { implicit driver =>
```
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/NotebookInteractionSpec.scala]: So we know Hail doesn't work -- but can we add a test to make sure Spark works with Py3?

I've tested using the "pi estimation" example on this page: https://spark.apache.org/examples.html

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/NotebooksListPage.scala [hunk 7]
```diff
@@ -7,6 +7,27 @@ import org.openqa.selenium.WebDriver
 
 import scala.util.Try
 
+sealed trait Kernel {
```
Comment (jmthibault79) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/NotebooksListPage.scala]: I like this.

File: jupyter-docker/Dockerfile (hunk 4 above)
Comment (jmthibault79) [reply jupyter-docker/Dockerfile]: The purpose of keeping it in the same line is to keep the layer small.  If the cleaning is done in a different step, the extra apt stuff remains in the earlier layer.

File: jupyter-docker/Dockerfile (hunk 5 above)
Comment (jmthibault79) [reply jupyter-docker/Dockerfile]: OK - hadn't seen this comment yet :)

**Comment by ansingh7115** [comment]
jenkins retest

**Comment by ansingh7115** [comment]
jenkins retest

**Comment by ansingh7115** [comment]
jenkins retest

**Comment by ansingh7115** [comment]
jenkins retest

File: jupyter-docker/Dockerfile [hunk 8]
```diff
@@ -127,23 +127,60 @@ RUN apt-get install -yq --no-install-recommends \
  && pip install bx-python \
  && pip install fastinterval \
  && pip install matplotlib-venn \
- && pip install "https://github.com/all-of-us/workbench/archive/pyclient-v0-1-rc3.zip#egg=aou_workbench_client&subdirectory=client/py"
+ && pip install "https://github.com/all-of-us/workbench/archive/pyclient-v0-1-rc3.zip#egg=aou_workbench_client&subdirectory=client/py" \
+ && apt-get clean \
+ && rm -rf /var/lib/apt/lists/*
+
```
Comment (ahaessly) [inline jupyter-docker/Dockerfile]: if we remove these 2 lines then we don't have to do the apt-get update below

**Comment by ansingh7115** [comment]
jenkins retest

**Pull request merged**